# Case

In this task not all data will be used, only **data in 2017**. So it is necessary to do filtering at the beginning. Besides that there are some players who make team transfers in the NBA transfer market so that there is **duplication of player data**. Therefore you can use the **df.drop_duplicates()** syntax to solve this to produce the same output as the trainer. **Delete columns that have as many missing values as the entire row of data**. Then you can do additional pre-processing if needed or you can immediately process the data.  

Column Descriptions: https://www.basketball-reference.com/about/glossary.html

# Data Import

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
raw_df = pd.read_csv("https://raw.githubusercontent.com/Syukrondzeko/Fellowship-5/main/Seasons_Stats.csv")

In [2]:
# Get only Year 2017 data
df = raw_df[raw_df["Year"]==2017].reset_index(drop=True)
df.head()

Unnamed: 0    Year        Player Pos   Age   Tm     G    GS      MP   PER  \
0       24096  2017.0  Alex Abrines  SG  23.0  OKC  68.0   6.0  1055.0  10.1   
1       24097  2017.0    Quincy Acy  PF  26.0  TOT  38.0   1.0   558.0  11.8   
2       24098  2017.0    Quincy Acy  PF  26.0  DAL   6.0   0.0    48.0  -1.4   
3       24099  2017.0    Quincy Acy  PF  26.0  BRK  32.0   1.0   510.0  13.1   
4       24100  2017.0  Steven Adams   C  23.0  OKC  80.0  80.0  2389.0  16.5   

     TS%   3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  blanl  \
0  0.560  0.724  0.144   1.9   7.1   4.5   5.5   1.7   0.6   8.3  15.9    NaN   
1  0.565  0.529  0.353   3.9  18.0  11.0   4.9   1.2   2.0   9.7  16.8    NaN   
2  0.355  0.412  0.176   4.6  15.2   9.7   0.0   0.0   0.0   9.8  20.0    NaN   
3  0.587  0.542  0.373   3.8  18.2  11.1   5.4   1.3   2.2   9.6  16.5    NaN   
4  0.589  0.002  0.392  13.0  15.5  14.2   5.4   1.8   2.6  16.0  16.2    NaN   

   OWS  DWS   WS  WS/48  blank2  OBPM  DBPM   BPM  VORP     FG    FGA    FG%  \
0  1.2  0.9  2.1  0.095     NaN  -0.3  -2.2  -2.5  -0.1  134.0  341.0  0.393   
1  0.5  0.5  0.9  0.082     NaN  -1.8  -1.2  -3.0  -0.1   70.0  170.0  0.412   
2 -0.2  0.0 -0.1 -0.133     NaN -10.1  -6.0 -16.2  -0.2    5.0   17.0  0.294   
3  0.6  0.5  1.1  0.102     NaN  -1.1  -0.7  -1.8   0.0   65.0  153.0  0.425   
4  3.3  3.1  6.4  0.130     NaN  -0.7   1.2   0.5   1.5  374.0  655.0  0.571   

     3P    3PA    3P%     2P    2PA    2P%   eFG%     FT    FTA    FT%    ORB  \
0  94.0  247.0  0.381   40.0   94.0  0.426  0.531   44.0   49.0  0.898   18.0   
1  37.0   90.0  0.411   33.0   80.0  0.413  0.521   45.0   60.0  0.750   20.0   
2   1.0    7.0  0.143    4.0   10.0  0.400  0.324    2.0    3.0  0.667    2.0   
3  36.0   83.0  0.434   29.0   70.0  0.414  0.542   43.0   57.0  0.754   18.0   
4   0.0    1.0  0.000  374.0  654.0  0.572  0.571  157.0  257.0  0.611  282.0   

     DRB    TRB   AST   STL   BLK    TOV     PF    PTS  
0   68.0   86.0  40.0  37.0   8.0   33.0  114.0  406.0  
1   95.0  115.0  18.0  14.0  15.0   21.0   67.0  222.0  
2    6.0    8.0   0.0   0.0   0.0    2.0    9.0   13.0  
3   89.0  107.0  18.0  14.0  15.0   19.0   58.0  209.0  
4  333.0  615.0  86.0  88.0  78.0  146.0  195.0  905.0

# Data Preprocessing

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595 entries, 0 to 594
Data columns (total 53 columns):
Unnamed: 0    595 non-null int64
Year          595 non-null float64
Player        595 non-null object
Pos           595 non-null object
Age           595 non-null float64
Tm            595 non-null object
G             595 non-null float64
GS            595 non-null float64
MP            595 non-null float64
PER           595 non-null float64
TS%           593 non-null float64
3PAr          593 non-null float64
FTr           593 non-null float64
ORB%          595 non-null float64
DRB%          595 non-null float64
TRB%          595 non-null float64
AST%          595 non-null float64
STL%          595 non-null float64
BLK%          595 non-null float64
TOV%          593 non-null float64
USG%          595 non-null float64
blanl         0 non-null float64
OWS           595 non-null float64
DWS           595 non-null float64
WS            595 non-null float64
WS/48         595 non-null 

In [4]:
# Check duplicate rows on Player column
len(df['Player'])-len(df['Player'].drop_duplicates())

109

In [5]:
# Drop duplicate names
df.drop_duplicates(subset="Player",inplace=True)

# Drop columns with 0 row
df = df.drop(['Unnamed: 0','blanl','blank2'],axis=1).reset_index(drop=True)

# Fill missing value with mean
df = df.fillna(df.mean())
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 486 entries, 0 to 485
Data columns (total 50 columns):
Year      486 non-null float64
Player    486 non-null object
Pos       486 non-null object
Age       486 non-null float64
Tm        486 non-null object
G         486 non-null float64
GS        486 non-null float64
MP        486 non-null float64
PER       486 non-null float64
TS%       486 non-null float64
3PAr      486 non-null float64
FTr       486 non-null float64
ORB%      486 non-null float64
DRB%      486 non-null float64
TRB%      486 non-null float64
AST%      486 non-null float64
STL%      486 non-null float64
BLK%      486 non-null float64
TOV%      486 non-null float64
USG%      486 non-null float64
OWS       486 non-null float64
DWS       486 non-null float64
WS        486 non-null float64
WS/48     486 non-null float64
OBPM      486 non-null float64
DBPM      486 non-null float64
BPM       486 non-null float64
VORP      486 non-null float64
FG        486 non-null float64

In [6]:
# Data statistics
df.describe()

Year         Age           G          GS           MP         PER  \
count   486.0  486.000000  486.000000  486.000000   486.000000  486.000000   
mean   2017.0   26.405350   53.783951   25.308642  1223.051440   13.020782   
std       0.0    4.345194   24.835638   28.715875   842.438143    5.762420   
min    2017.0   19.000000    1.000000    0.000000     1.000000  -17.600000   
25%    2017.0   23.000000   35.250000    1.000000   449.500000    9.800000   
50%    2017.0   26.000000   62.500000   11.000000  1197.500000   12.800000   
75%    2017.0   29.000000   75.000000   49.750000  1942.250000   15.800000   
max    2017.0   40.000000   82.000000   82.000000  3048.000000   31.500000   

              TS%        3PAr         FTr        ORB%        DRB%        TRB%  \
count  486.000000  486.000000  486.000000  486.000000  486.000000  486.000000   
mean     0.526944    0.314058    0.270179    5.067901   15.211934   10.145267   
std      0.089679    0.209531    0.180101    4.246585    7.337450    5.201848   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.502250    0.156000    0.165000    1.900000   10.325000    6.225000   
50%      0.537000    0.323500    0.237000    3.550000   14.050000    8.950000   
75%      0.575750    0.454250    0.339000    7.700000   18.875000   13.100000   
max      0.799000    1.000000    2.000000   26.300000  100.000000   56.400000   

             AST%        STL%        BLK%        TOV%        USG%         OWS  \
count  486.000000  486.000000  486.000000  486.000000  486.000000  486.000000   
mean    13.067695    1.545062    1.664403   12.809072   18.608848    1.335597   
std      9.172849    0.923243    1.690613    5.326882    5.597041    2.081620   
min      0.000000    0.000000    0.000000    0.000000    0.000000   -1.700000   
25%      6.525000    1.100000    0.500000    9.625000   14.600000    0.000000   
50%     10.100000    1.400000    1.200000   12.450000   18.100000    0.600000   
75%     17.575000    1.900000    2.400000   15.400000   21.500000    2.000000   
max     57.300000   11.100000   17.300000   43.600000   41.700000   11.500000   

              DWS          WS       WS/48        OBPM        DBPM         BPM  \
count  486.000000  486.000000  486.000000  486.000000  486.000000  486.000000   
mean     1.243416    2.580247    0.077249   -1.397531   -0.304733   -1.701852   
std      1.085874    2.898265    0.082869    3.396019    2.033137    4.220521   
min      0.000000   -0.800000   -0.473000  -22.400000   -7.100000  -26.900000   
25%      0.400000    0.400000    0.041000   -2.900000   -1.600000   -3.600000   
50%      1.000000    1.800000    0.082000   -1.300000   -0.300000   -1.500000   
75%      1.875000    3.775000    0.116750    0.175000    0.900000    0.375000   
max      6.000000   15.000000    0.480000   11.800000   12.000000   15.600000   

             VORP          FG          FGA         FG%          3P  \
count  486.000000  486.000000   486.000000  486.000000  486.000000   
mean     0.616667  197.656379   432.337449    0.441268   48.864198   
std      1.454814  174.667670   371.858588    0.098991   56.011800   
min     -1.400000    0.000000     0.000000    0.000000    0.000000   
25%     -0.100000   55.000000   120.250000    0.399000    3.000000   
50%      0.100000  160.500000   361.000000    0.442000   28.500000   
75%      0.900000  289.750000   633.750000    0.484750   78.000000   
max     12.400000  824.000000  1941.000000    1.000000  324.000000   

              3PA         3P%          2P          2PA         2P%  \
count  486.000000  486.000000  486.000000   486.000000  486.000000   
mean   136.676955    0.299342  148.792181   295.660494    0.482702   
std    147.204832    0.126221  142.477490   275.717392    0.109121   
min      0.000000    0.000000    0.000000     0.000000    0.000000   
25%     10.000000    0.275250   38.000000    82.000000    0.445250   
50%     94.000000    0.331500  109.000000   228.000000    0.48900

# Tasks

## Who is the youngest and oldest player in the NBA in 2017 for each team (Tm) ?

In [7]:
# Get max and min age for each team
team_max_min_age = df.groupby('Tm')['Age'].agg(MaxAge=('Max Age', 'max'), MinAge=('Min Age', 'min')).reset_index()
team_max_min_age

Tm  MaxAge  MinAge
0   ATL    32.0    22.0
1   BOS    31.0    20.0
2   BRK    36.0    21.0
3   CHI    35.0    21.0
4   CHO    31.0    21.0
5   CLE    38.0    21.0
6   DAL    38.0    21.0
7   DEN    36.0    19.0
8   DET    34.0    20.0
9   GSW    36.0    20.0
10  HOU    34.0    20.0
11  IND    32.0    20.0
12  LAC    39.0    19.0
13  LAL    37.0    19.0
14  MEM    40.0    20.0
15  MIA    36.0    20.0
16  MIL    39.0    19.0
17  MIN    34.0    20.0
18  NOP    33.0    20.0
19  NYK    32.0    21.0
20  OKC    36.0    20.0
21  ORL    32.0    20.0
22  PHI    32.0    21.0
23  PHO    34.0    19.0
24  POR    28.0    21.0
25  SAC    31.0    19.0
26  SAS    39.0    20.0
27  TOR    30.0    21.0
28  TOT    36.0    21.0
29  UTA    35.0    21.0
30  WAS    32.0    21.0

In [8]:
# Get oldest player name from each team
team_oldest = []
for i in range(len(team_max_min_age)):
    team_oldest.append(df.loc[(df["Age"]==team_max_min_age["MaxAge"].values[i]) & 
                              (df["Tm"]==team_max_min_age["Tm"].values[i]),
                              "Player"].values[0])
    
# Get youngest player name from each team
team_youngest = []
for i in range(len(team_max_min_age)):
    team_youngest.append(df.loc[(df["Age"]==team_max_min_age["MinAge"].values[i]) & 
                              (df["Tm"]==team_max_min_age["Tm"].values[i]),
                              "Player"].values[0])

In [9]:
# Assign it to the previous dataframe
team_max_min_age["Player Oldest"] = team_oldest
team_max_min_age["Player Youngest"] = team_youngest
team_max_min_age

Tm  MaxAge  MinAge    Player Oldest          Player Youngest
0   ATL    32.0    22.0        Gary Neal          DeAndre' Bembry
1   BOS    31.0    20.0     Gerald Green             Jaylen Brown
2   BRK    36.0    21.0       Luis Scola         Isaiah Whitehead
3   CHI    35.0    21.0      Dwyane Wade             Bobby Portis
4   CHO    31.0    21.0    Brian Roberts           Christian Wood
5   CLE    38.0    21.0   Chris Andersen               Kay Felder
6   DAL    38.0    21.0    Dirk Nowitzki               Ben Bentil
7   DEN    36.0    19.0      Mike Miller             Jamal Murray
8   DET    34.0    20.0       Beno Udrih           Henry Ellenson
9   GSW    36.0    20.0       David West             Kevon Looney
10  HOU    34.0    20.0     Nene Hilario           Chinanu Onuaku
11  IND    32.0    20.0     Aaron Brooks             Myles Turner
12  LAC    39.0    19.0      Paul Pierce            Diamond Stone
13  LAL    37.0    19.0      Metta World           Brandon Ingram
14  MEM    40.0    20.0     Vince Carter             Wade Baldwin
15  MIA    36.0    20.0    Udonis Haslem          Justise Winslow
16  MIL    39.0    19.0      Jason Terry               Thon Maker
17  MIN    34.0    20.0       John Lucas               Tyus Jones
18  NOP    33.0    20.0     Jarrett Jack            Cheick Diallo
19  NYK    32.0    21.0  Carmelo Anthony       Kristaps Porzingis
20  OKC    36.0    20.0    Nick Collison         Domantas Sabonis
21  ORL    32.0    20.0      C.J. Watson        Stephen Zimmerman
22  PHI    32.0    21.0   Tiago Splitter  Timothe Luwawu-Cabarrot
23  PHO    34.0    19.0  Leandro Barbosa            Dragan Bender
24  POR    28.0    21.0      Evan Turner              Noah Vonleh
25  SAC    31.0    19.0    Arron Afflalo     Georgios Papagiannis
26  SAS    39.0    20.0    Manu Ginobili          Dejounte Murray
27  TOR    30.0    21.0  DeMarre Carroll            Bruno Caboclo
28  TOT    36.0    21.0      Matt Barnes         Chris McCullough
29  UTA    35.0    21.0      Joe Johnson               Dante Exum
30  WAS    32.0    21.0    Marcin Gortat              Kelly Oubre

## Which player has the most minutes played (MP) in each position (Pos)?

In [10]:
# Get max minutes player in each position
df_pos_max_mp = df.groupby('Pos')['MP'].agg(MaxMP=('Max MP', 'max')).reset_index()
df_pos_max_mp

Pos   MaxMP
0     C  3030.0
1    PF  2803.0
2  PF-C   980.0
3    PG  2947.0
4    SF  3048.0
5    SG  2796.0

In [11]:
# Get player name with max minutes in each position
pos_most_mp = []
for i in range(len(df_pos_max_mp)):
    pos_most_mp.append(df.loc[(df["MP"]==df_pos_max_mp["MaxMP"].values[i]) & 
                              (df["Pos"]==df_pos_max_mp["Pos"].values[i]),
                              "Player"].values[0])
pos_most_mp

['Karl-Anthony Towns',
 'Harrison Barnes',
 'Joffrey Lauvergne',
 'James Harden',
 'Andrew Wiggins',
 'C.J. McCollum']

In [12]:
# Assign list to the dataframe
df_pos_max_mp["Player"] = pos_most_mp
df_pos_max_mp

Pos   MaxMP              Player
0     C  3030.0  Karl-Anthony Towns
1    PF  2803.0     Harrison Barnes
2  PF-C   980.0   Joffrey Lauvergne
3    PG  2947.0        James Harden
4    SF  3048.0      Andrew Wiggins
5    SG  2796.0       C.J. McCollum

These are the player names with the most Minutes Played. These players should have the most experience until 2017 NBA.

## Which team has the highest average total rebound percentage (TRB%), assist percentage (AST%), steal percentage (STL%), and block percentage (BLK%)?

In [13]:
# Get metrics
metrics = ["TRB%","AST%","STL%","BLK%"]

In [14]:
# Function to find the team with best average metrics
def get_avg_metrics(metrics):
    var = df.groupby("Tm")[metrics].mean().reset_index().sort_values(metrics,ascending=False).values[0]
    print("Team with the highest",metrics,"is",var[0],"with",round(var[1],2),metrics)

In [15]:
# Find the team with the highest above metrics
for metric in metrics:
    get_avg_metrics(metric)

Team with the highest TRB% is WAS with 13.45 TRB%
Team with the highest AST% is DEN with 15.86 AST%
Team with the highest STL% is MIN with 2.37 STL%
Team with the highest BLK% is GSW with 2.74 BLK%


## Who is the best player in your opinion based on his record stats? note: you can refer to variables point (PTS), assists, rebounds, or anything else. A combination of several variables would be nice.

The Player Efficiency Rating (PER) is a per-minute rating developed by ESPN.com columnist John Hollinger. In John's words, "The PER sums up all a player's positive accomplishments, subtracts the negative accomplishments, and returns a per-minute rating of a player's performance." It appears from his books that John's database only goes back to the 1988-89 season.  

All calculations begin with what I am calling unadjusted PER (uPER). The formula is:

     uPER = (1 / MP) *
     [ 3P
     + (2/3) * AST
     + (2 - factor * (team_AST / team_FG)) * FG
     + (FT *0.5 * (1 + (1 - (team_AST / team_FG)) + (2/3) * (team_AST / team_FG)))
     - VOP * TOV
     - VOP * DRB% * (FGA - FG)
     - VOP * 0.44 * (0.44 + (0.56 * DRB%)) * (FTA - FT)
     + VOP * (1 - DRB%) * (TRB - ORB)
     + VOP * DRB% * ORB
     + VOP * STL
     + VOP * DRB% * BLK
     - PF * ((lg_FT / lg_PF) - 0.44 * (lg_FTA / lg_PF) * VOP) ]  

Most of the terms in the formula above should be clear, but let me define the less obvious ones:

    factor = (2 / 3) - (0.5 * (lg_AST / lg_FG)) / (2 * (lg_FG / lg_FT))
    VOP    = lg_PTS / (lg_FGA - lg_ORB + lg_TOV + 0.44 * lg_FTA)
    DRB%   = (lg_TRB - lg_ORB) / lg_TRB

The calcuation of uPER obviously depends on these statistics, so here are my solutions for years when the data are missing:

- Zero out three-point field goals, turnovers, blocked shots, and steals.
- Set the league value of possession (VOP) equal to 1.
- Set the defensive rebound percentage (DRB%) equal to 0.7.
- Set player offensive rebounds (ORB) equal to 0.3 * TRB.  

Some of these solutions may not be elegant, but I think they are reasonable. After uPER is calculated, an adjustment must be made for the team's pace. The pace adjustment is:

    pace adjustment = lg_Pace / team_Pace
    
League and team pace factors cannot be computed for seasons prior to 1973-74, so I estimate the above using:

    estimated pace adjustment = 2 * lg_PPG / (team_PPG + opp_PPG)

Now the pace adjustment is made to uPER (I will call this aPER):

    aPER = (pace adjustment) * uPER
The final step is to standardize aPER. First, calculate league average aPER (lg_aPER) using player minutes played as the weights. Then, do the following:

    PER = aPER * (15 / lg_aPER)

Link to PER formula explanation : https://www.basketball-reference.com/about/per.html

The PER formula itself is complex to calculate. It involves a lot of variables to get the estimated player efficiency. Therefore, This analysis would only use PER column.

In [16]:
# Get a player name and statuses with the highest PER
df[df["PER"]==df["PER"].max()][["Player","PER","G","PTS","MP"]]

Player   PER    G  PTS   MP
413  Jarnell Stokes  31.5  2.0  3.0  7.0

According to the above output, Jamell Stokes has the highest PER. But, if we take a look at his PTS, and G columns, the value looks very low. I assume Jamell Stokes is a new player with great potentiality. Therefore, Let's take a look on how the top 10 players with the highest PER.

In [17]:
# Get top 10 player and statuses with the highest PER
df.sort_values('PER',ascending=False).head(10)[["Player","PER","G","PTS","MP"]]

Player   PER     G     PTS      MP
413     Jarnell Stokes  31.5   2.0     3.0     7.0
216  Demetrius Jackson  30.8   5.0    10.0    17.0
457  Russell Westbrook  30.6  81.0  2558.0  2802.0
280   Boban Marjanovic  29.6  35.0   191.0   293.0
118       Kevin Durant  27.6  62.0  1555.0  2070.0
260      Kawhi Leonard  27.5  74.0  1888.0  2474.0
99       Anthony Davis  27.5  75.0  2099.0  2708.0
171       James Harden  27.3  81.0  2356.0  2947.0
219       LeBron James  27.0  74.0  1954.0  2794.0
423      Isaiah Thomas  26.5  76.0  2199.0  2569.0

After Jamell Stokes, Demetrius Jackson has the second highest PER. But, both of their G and PTS values looks very low. They need to participate more in NBA to make the PER value much more credible.  
If we take a look more at the data, Russell Westbrook gained the 3rd highest PER. Other than that, he gained a high PTS and G values, which is 2558 and 81 respectively.  
Therefore, according to this analysis, **Russell Westbrook** is the best player.

## Which team has the best average stat record of their players? Note: you can refer to points, assists, rebounds, or anything else. A combination of several variables would be nice

For this Analysis, I would like to use TS%, FT%, eFG%, DRB%, and BLK% columns (For the column details, please take a look at the Glossary link at the top page of this notebook). These columns would be averaged for each player. And then, player's averaged score would be grouped by team to find the best team.

In [18]:
from sklearn.preprocessing import MinMaxScaler

#Create scaler object
scaler = MinMaxScaler()
metrics = ["TS%","FT%","eFG%","DRB%","BLK%"]

scaled_df = df.copy()
df[metrics]

TS%    FT%   eFG%  DRB%  BLK%
0    0.560  0.898  0.531   7.1   0.6
1    0.565  0.750  0.521  18.0   2.0
2    0.589  0.611  0.571  15.5   2.6
3    0.559  0.892  0.514   8.4   0.4
4    0.529  0.725  0.500  23.8   3.1
..     ...    ...    ...   ...   ...
481  0.604  0.679  0.571  17.3   3.0
482  0.508  0.564  0.494  17.0   3.3
483  0.346  0.600  0.323  24.9   3.7
484  0.503  0.775  0.473  14.2   1.5
485  0.547  0.653  0.529  21.9   4.4

[486 rows x 5 columns]

In [19]:
# Rescale the values
scaled_df[metrics] = scaler.fit_transform(scaled_df[metrics])
scaled_df[metrics]

TS%    FT%   eFG%   DRB%      BLK%
0    0.700876  0.898  0.531  0.071  0.034682
1    0.707134  0.750  0.521  0.180  0.115607
2    0.737171  0.611  0.571  0.155  0.150289
3    0.699625  0.892  0.514  0.084  0.023121
4    0.662078  0.725  0.500  0.238  0.179191
..        ...    ...    ...    ...       ...
481  0.755945  0.679  0.571  0.173  0.173410
482  0.635795  0.564  0.494  0.170  0.190751
483  0.433041  0.600  0.323  0.249  0.213873
484  0.629537  0.775  0.473  0.142  0.086705
485  0.684606  0.653  0.529  0.219  0.254335

[486 rows x 5 columns]

In [20]:
# Find average status
additional_metrics = metrics.copy()
additional_metrics.append("avgStat%")

scaled_df["avgStat%"] = scaled_df[metrics].mean(axis=1)
scaled_df[additional_metrics]

TS%    FT%   eFG%   DRB%      BLK%  avgStat%
0    0.700876  0.898  0.531  0.071  0.034682  0.447112
1    0.707134  0.750  0.521  0.180  0.115607  0.454748
2    0.737171  0.611  0.571  0.155  0.150289  0.444892
3    0.699625  0.892  0.514  0.084  0.023121  0.442549
4    0.662078  0.725  0.500  0.238  0.179191  0.460854
..        ...    ...    ...    ...       ...       ...
481  0.755945  0.679  0.571  0.173  0.173410  0.470471
482  0.635795  0.564  0.494  0.170  0.190751  0.410909
483  0.433041  0.600  0.323  0.249  0.213873  0.363783
484  0.629537  0.775  0.473  0.142  0.086705  0.421248
485  0.684606  0.653  0.529  0.219  0.254335  0.467988

[486 rows x 6 columns]

In [21]:
# Sort team names by the highest average status
scaled_df.groupby("Tm")["avgStat%"].mean().reset_index().sort_values("avgStat%",ascending=False)

Tm  avgStat%
26  SAS  0.454016
9   GSW  0.452604
7   DEN  0.446155
12  LAC  0.443542
25  SAC  0.442405
29  UTA  0.441357
1   BOS  0.438883
22  PHI  0.437931
30  WAS  0.437288
10  HOU  0.436645
18  NOP  0.435301
16  MIL  0.434859
5   CLE  0.434558
27  TOR  0.432800
14  MEM  0.430883
28  TOT  0.430172
19  NYK  0.429128
2   BRK  0.429059
11  IND  0.427939
24  POR  0.425983
17  MIN  0.423700
15  MIA  0.420625
13  LAL  0.418484
8   DET  0.415103
23  PHO  0.413968
4   CHO  0.412823
20  OKC  0.411053
6   DAL  0.410647
3   CHI  0.407021
0   ATL  0.403568
21  ORL  0.399683

It looks like SAS team has the highest average total stats based on the average of TS%, FT%, eFG%, DRB%, and BLK% columns of each player. Therefore, they're the best team in NBA.